### ISSUE TITLE
FP8 E4M3 decode uses a single per-tensor weight absmax — 9.6x the logits error of every other GEMM route

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#2810]
`get_or_cache_fp8_weight` puts an entire `[n x k]` weight through a **single per-tensor absmax** before quantizing to E4M3's 3 mantissa bits. On the decode path that costs **9.6x** the logits error of every other GEMM route, and it is enough to move a greedy argmax.

Measured on RTX 4090 sm_89 while closing #2753, `qwen2.5-coder-1.5b-instruct-q4_k_m`, ordered streams, `APR_PARITY_PROBE=1` at m=1 forced onto the batched path (`APR_FORCE_BATCHED_PATH=1 CUBLAS_GEMM_THRESHOLD=1`), **all four probe controls passing** (SELF 1.000000, SENSITIVE and PERTURBED diverging), one binary, one env var each:

| `APR_DECODE_GEMM` | route | ORACLE cosine | max\|d\| | answer |
|---|---|---|---|---|
| `cublas` (default) | FP8 E4M3 | 0.998774 | **0.8655** | "Compilers are a crucial component of computer programming" |
| `nolmhead` | FP8 except LM head | 0.999184 | 0.6662 | "Compilers are software programs that convert" |
| `gemv` | DP4A / per-row GEMV | 0.999983 | **0.0905** | "Compilers are software programs that take" |
| — | m=1 fast path (reference) | — | — | "Compilers are software programs that translate" |

on logits of absmax ~14.

## Why it is weight-side, not activation-side

**At m = 1 the activation scale is already per-row** — there is one row — and the 0.8655 reproduces there. So the residual is in the weight quantization, not in the per-tensor activation absmax that `cublas_prefill_fp8_gemm` computes per call.

## Why the obvious workaround is the wrong fix

Routing decode off cuBLAS entirely removes the error but costs aggregate throughput (token-normalised, 120 tokens/request, median of 3, same binary):

| route | c=1 | c=4 | c=8 |
|---|---|---|---|
| `cublas` | 196.9 | 457.2 | 900.8 |
| `nolmhead` | 193.2 | 400.8 (0.877x) | 724.0 (0.804x) |
| `gemv` | 200.2 | 288.8 (**0.632x**) | 351.6 (**0.390x**) |

37-61% of aggregate throughput for a near-tie flip, when `nolmhead` buys only 23% of the error for 12-20% of the throughput. `APR_DECODE_GEMM` therefore ships **default `cublas`** (#2809) as a measurement lever only.

## The repair

**Per-output-channel scales.** `D[j, i] = sum_k W[j,k] * A[i,k]`, so a per-`j` weight scale is a per-output-feature multiply on the result. Concretely:

1. `get_or_cache_fp8_weight`: per-row absmax over `k` -> a device vector of `n` floats; quantize each row by its own scale (today: one `gpu_absmax` over `n*k` and one scalar `quant_scale`).
2. Store the per-row dequant scales as a device buffer keyed by `weight_ptr` (today: `fp8_weight_scales: HashMap<u64, f32>`).
3. `cublas_prefill_fp8_gemm`: `alpha = 1.0` instead of the scalar `weight_dequant`, and `convert_f16_to_f32_act_scaled` multiplies element `idx` by `act_dequant * w_scale[idx % n]` — the output is `[m x n]` with `ldd = n`, so `idx % n` is the output feature.

This keeps the tensor-core GEMM and its throughput. It also improves **prefill**, which uses the same cache, so the m=1 reference itself may shift — that is expected and should be measured, not assumed.

## How to know it worked

The ladder above is the acceptance test, and it is already runnable:

```
APR_STREAM_LEGACY not needed post-#2776
APR_FORCE_BATCHED_PATH=1 APR_PARITY_PROBE=1 CUBLAS_GEMM_THRESHOLD=1 \
  apr serve run <gguf> --gpu --port <p> --context-length 4096
```

Target: FP8's `max|d|` at or below the FP16/GEMV floor of ~0.09-0.15 with `agg c=8` within noise of 900.8. If per-channel scaling does **not** close it, that is itself the finding — it would mean 3 mantissa bits are the floor and the answer is to route decode off FP8 and pay the throughput, which is the decision #2809 declined to make without this datum.

## Also surfaced while measuring this

- **`FP8_DECODE=0` does not control FP8 on the decode path.** `cublas_prefill_gemm` selects FP8 on `gpu_profile.fp8_prefill`, so once decode reaches it at `m >= cublas_gemm_threshold()` the route is FP8 regardless. Measured: at the default threshold, `FP8_DECODE=0` leaves the c=4 and c=8 output **byte-identical**. #2789's table saw it "restore m=6/8/16" because that row also set `CUBLAS_GEMM_THRESHOLD=32`, which is what actually moved the route. The variable only guards the separate `m >= 5` shortcut in `batched_gemv_or_gemm`.
- **`CUBLAS_GEMM_THRESHOLD=32` with `FP8_DECODE=0` makes the m=1 REFERENCE garbage** (`Lambda*****` / `**0*000...`), because prefill then falls into `batched_gemv_with_fallback`'s per-sequence loop (#2761). Pre-existing, reproduced on unmodified main; the parity probe correctly returns 2 UNMEASURABLE rather than a verdict.

Refs #2753, #2809, #2789, #2761

**Comment by noahgift** [comment]
0.67 lane — attaches to PP-066 spec §6 row **W-H**.

Per-output-channel FP8 scales are a kernel numerics repair on the cuBLAS decode/prefill route; PP-066 keeps APR_DECODE_GEMM=cublas as a measurement lever (#2809) and sends kernel work to 0.67. The 9.6× logits error is the mechanism behind the m≥4 family parting that W-H's margin instrument classifies — same row as #2770.

_PP-066 triage 2026-09-05 (issue #2873); ledger `docs/audits/pp-066-triage-2026-09-05.md`._

**Comment by noahgift** [comment]
triage 2026-09-23 (aprender-dd, standing triage owner): milestoned to 0.73.0, theme llama.cpp Parity (performance, kernels, parity benches); it moved out of `backlog` (the 2026-09-17 overflow bucket, not a release).

**Comment by noahgift** [comment]
Closed by paiml-implement — folded into epic #3999 (triage evidence/triage-1717.md)
